### ISSUE TITLE
How to insert 2 different type of relations?

### ISSUE THREAD

**Issue by fjook** [Vincit/objection.js#1813]
TL;DR - How do I insert a `user`'s `product` (OneToMany) which has multiple `categories` (ManyToMany)?

---

Hi! So I have these four tables:
```js
// users
...
      table.increments('id').primary()
      table.string('name')
...

// categories
...
      table.increments('id').primary()

      table
        .integer('parent_id')
        .unsigned()
        .references('id')
        .inTable('categories')
        .onDelete('SET NULL')
        .index()

      table.string('name').notNullable()
...

// products
...
      table.increments('id').primary()

      table
        .integer('user_id')
        .unsigned()
        .references('id')
        .inTable('users')
        .onDelete('CASCADE')
        .index()

      table.text('name').notNullable()
...

// product_categories
...
      table.increments('id').primary()

      table
        .integer('product_id')
        .unsigned()
        .references('id')
        .inTable('products')
        .onDelete('CASCADE')
        .index()

      table
        .integer('category_id')
        .unsigned()
        .references('id')
        .inTable('categories')
        .onDelete('SET NULL')
        .index()
...
```

These are models:
```ts
// user.ts
...
  static relationMappings = (): any => ({
    products: {
      relation: Model.HasManyRelation,
      modelClass: Product,
      join: {
        from: 'users.id',
        to: 'products.user_id',
      },
    },
  })
...

// category.ts
...
  static relationMappings = (): any => ({
    parent: {
      relation: Model.BelongsToOneRelation,
      modelClass: Category,
      join: {
        from: 'categories.parent_id',
        to: 'categories.id',
      },
    },

    children: {
      relation: Model.HasManyRelation,
      modelClass: Category,
      join: {
        from: 'categories.id',
        to: 'categories.parent_id',
      },
    },

    products: {
      relation: Model.ManyToManyRelation,
      modelClass: Product,
      join: {
        from: 'categories.id',
        through: {
          from: 'product_categories.category_id',
          to: 'product_categories.product_id',
        },
        to: 'products.id',
      },
    },
  })
...

// product.ts
...
  static relationMappings = (): any => ({
    user: {
      relation: Model.BelongsToOneRelation,
      modelClass: User,
      join: {
        from: 'products.user_id',
        to: 'users.id',
      },
    },

    categories: {
      relation: Model.ManyToManyRelation,
      modelClass: Category,
      join: {
        from: 'products.id',
        through: {
          from: 'product_categories.product_id',
          to: 'product_categories.category_id',
        },
        to: 'categories.id',
      },
    },
  })
...
```

And this is the route and controller for `user`:
```ts
// Took reference from Objection's example koa-ts
router.post('/users/:id/products', checkAuth, async (ctx: Context) => {
  const product = await User.relatedQuery('products')
    .for(ctx.params.id)
    .insert(ctx.request.body)
  return (ctx.body = product)
})
```

Sample POST request body of adding user product at `http://localhost:8000/users/1/products`:

```json
{
  "name": "Product One",
  "categories": [1, 2, 3]  // This can be changed
}
```

Should I map those `categories` values with `categories.map(id => Categories.relateQuery('products').for(id).relate(newProduct.id))`? Pretty sure there should be a right and better way to do it.

Please enlighten me.

---

Stack: TypeScript (v3.9.7), Koa (v2.13.0), Objection (v2.2.2), Knex (v0.21.2), Pg (v8.3.0), Node (12.18.3), Postgres (v12.3)

**Comment by koskimas** [comment]
See the insertGraph method.

**Comment by fjook** [comment]
Thank you so much.
